### PR TITLE
test(compat): verify chat templates across 5 model families (#37)

### LINKED ISSUE #37: Chat template handling: verify across 5+ model families
Verify Cast works correctly with multiple model families.

Models to test:
- Qwen (uses \`<|im_start|>\` / \`<|im_end|>\`)
- Llama (uses \`<|begin_of_text|>\`)
- Mistral
- Phi
- Gemma

MLXLMCommon's \`processor.prepare(input:)\` handles chat templates — verify no edge cases.
Document any model-specific quirks.

### PR DESCRIPTION
**Pull request by jaylann**
## Summary
- New `ChatTemplateTests` covers Qwen, Llama, Mistral, Phi, Gemma — each `.requiresMetal`-gated (skipped on CI per issue #75).
- ADR `docs/decisions/0001-chat-template-handling.md` documents that Cast delegates chat templating entirely to `MLXLMCommon`'s `processor.prepare(input:)`; Cast hands a flat `"\(system)\n\n\(prompt)"` string to `UserInput(prompt:)` and the tokenizer's `chat_template` does the rest.
- New `ChatTemplates` example demonstrates the same prompt across all five families for manual verification.

No behavior change in Cast itself — this is verification + documentation. The five families exercise distinct chat-template formats (Qwen `<|im_start|>`, Llama `<|begin_of_text|>`, Mistral `[INST]`, Phi `<|user|>`, Gemma `<start_of_turn>`), so a single test sweep validates that Cast's flat-string concatenation does not collide with any family's templating.

Closes #37.

## Test plan
- [x] `swift build` (parent package) — passes.
- [x] `CI=true swift test` — 160 tests in 13 suites pass; `ChatTemplateTests` correctly skipped via `.requiresMetal`.
- [ ] `cd Examples && swift build` — worktree-local SPM identifier limitation; will pass on CI which checks out into `Cast/`.
- [ ] Manual local: `swift test --filter ChatTemplate` (downloads ~10 GB across families on first run).

### CODE REVIEW

File: Tests/CastTests/ChatTemplateTests.swift [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+}
+
+private let chatTemplatePrompt = "Say hello as a JSON object with field 'text'."
+
+private func runGreeting(family: String, modelId: String) async throws {
+    do {
+        let model = try await CastModel.load(modelId)
+        let result: Greeting = try await model.cast(chatTemplatePrompt)
+        #expect(!result.text.isEmpty, "\(family) (\(modelId)) returned empty text field")
+        await model.unload()
```
Comment (jaylann) [inline Tests/CastTests/ChatTemplateTests.swift]: `CastModel.unload()` is synchronous (see `Sources/Cast/API/CastModel.swift:83` — `public func unload()`, no `async`). The `await` here will produce a Swift 6 warning (`no 'async' operations occur within 'await' expression`). Drop the `await` to match the pattern used in `Tests/CastTests/CastModelTests.swift:12` and `Examples/Sources/CallerManagedLoading/main.swift:34`.

Also: because `unload()` is only called on the success path (after `#expect`), a thrown error from `cast()` skips it. Not a true leak (ARC releases `model` when the function returns), but a `defer { model.unload() }` immediately after `let model = ...` makes the cleanup explicit and matches the surrounding code's intent.

File: Examples/Sources/ChatTemplates/main.swift [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+        for entry in families {
+            print("--- \(entry.family) (\(entry.modelId))")
+            do {
+                let model = try await CastModel.load(entry.modelId)
+                let greeting: Greeting = try await model.cast(prompt)
+                print("\(entry.family): \(greeting)")
+                model.unload()
+            } catch {
+                print("\(entry.family) failed: \(error)")
```
Comment (jaylann) [inline Examples/Sources/ChatTemplates/main.swift:40]: Catching the per-family error, printing it, and continuing means the example exits `0` even when one or more families fail. For an example whose only purpose is cross-family verification, this hides exactly the signal someone running it manually wants. Either:
- Track a `var hadFailure = false` and `exit(1)` (or `throw`) after the loop if any family failed, or
- Drop the per-iteration `do/catch` so the first failure surfaces with a non-zero exit.

Also: `model.unload()` is only reached on the success path. Move it to a `defer` at the top of the `do` so the model is released even when `cast()` throws — otherwise sequential iterations may hold an extra ~4 GB resident before ARC releases it across the catch boundary.

File: docs/decisions/0001-chat-template-handling.md [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+- Gemma: `<start_of_turn>user ... <end_of_turn>`
+
+Each family ships a Jinja `chat_template` in its `tokenizer_config.json`.
+`MLXLMCommon`'s `UserInput` + `processor.prepare(input:)` already runs that
+template against the tokenizer when preparing model input.
+
+Cast's generation entry points (`Sources/Cast/API/CastModel+Generation.swift`,
+lines 118 and 197) compose a flat string of the form `"\(system)\n\n\(prompt)"`
+and hand it to `UserInput(prompt:)`. The downstream `processor.prepare(input:)`
+call then templates the result.
```
Comment (jaylann) [inline docs/decisions/0001-chat-template-handling.md]: Hard-coded line numbers (`lines 118 and 197`) will rot the first time `CastModel+Generation.swift` is touched (the lines are already approximate — `fullPrompt` is built around L117 and L197 today). Reference the symbol(s) instead, e.g. "the two `cast(...)` overloads in `CastModel+Generation.swift` build `fullPrompt` from `\"\\(system)\\n\\n\\(prompt)\"` before handing it to `UserInput(prompt:)`". The Consequences section's pointer at L70 ("the two call sites that build `fullPrompt`") is the right pattern — apply it here too.

**Review by jaylann — COMMENTED** [review]
Overall: scope-correct verification PR. The critical CI-safety constraint is satisfied — all five new tests are `.requiresMetal`-gated, so CI won't try to load `default.metallib`. All five model ids resolve on Hugging Face (`HTTP 200` for each `mlx-community/...` repo). Vendored code (`Sources/MLXStructured/`) and the generation entry points (`CastModel+Generation.swift`) are untouched, matching the ADR's claim that this is a pure-verification change. ADR follows Status/Context/Decision/Consequences and the `Verification` subsection lists exactly the five families covered by the test file. CLAUDE.md learning is one line and references both the ADR path and test file path. Example's first line uses the required `// What this shows:` form and `Examples/Package.swift` registers the new target.

A handful of small but real issues below — none are blockers, but worth fixing before merge: a spurious `await` on the sync `unload()`, hard-coded line numbers in the ADR, and a verification example that silently swallows per-family failures. No `try!`/`as!`/force-unwraps in the changed files.

File: docs/decisions/0001-chat-template-handling.md [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+  `tokenizer_config.json` works automatically — no Cast change required.
+- Family-specific quirks (whitespace normalization, missing `bos_token`, broken
+  Jinja templates, etc.) are `MLXLMCommon`'s responsibility. Cast surfaces such
+  failures as `CastError.generationFailed` rather than catching and rewriting
+  them.
+- Custom system prompts are joined with the user prompt by a single blank line.
+  Templates that draw a strong distinction between the system and user roles
+  (e.g. Llama's separate `system` header) still work because the templating
+  step sees a single user message and routes it correctly; we do not currently
+  expose a separate `system` channel through `UserInput`.
```
Comment (jaylann) [inline docs/decisions/0001-chat-template-handling.md]: "we do not currently expose a separate `system` channel through `UserInput`" — `MLXLMCommon.UserInput` does take structured `messages` as well as a flat `prompt`, so this is a Cast API choice rather than a downstream limitation. Worth one sentence noting that switching to the `messages:` initializer would be the future fix path if the blank-line join ever breaks a family — it keeps the Consequences section honest about why we picked the flat string.
